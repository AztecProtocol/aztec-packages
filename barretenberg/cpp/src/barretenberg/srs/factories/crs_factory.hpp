#pragma once
#include "barretenberg/common/mem.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"
#include "barretenberg/ecc/curves/bn254/g2.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
#include <cstddef>

namespace bb::pairing {
struct miller_lines;
} // namespace bb::pairing

namespace bb::srs::factories {

/**
 * A prover crs representation.
 */
template <typename Curve> class ProverCrs {
  public:
    virtual ~ProverCrs() = default;
    ;

    /**
     *  @brief Returns the monomial points in a form to be consumed by scalar_multiplication pippenger algorithm.
     */
    virtual typename Curve::AffineElement* get_monomial_points() = 0;
    virtual size_t get_monomial_size() const = 0;
};

template <typename Curve> class VerifierCrs {
  public:
    virtual ~VerifierCrs() = default;
};
template <> class VerifierCrs<curve::BN254> {
    using Curve = curve::BN254;

  public:
    virtual Curve::G2AffineElement get_g2x() const = 0;
    /**
     * @brief As the G_2 element of the CRS is fixed, we can precompute the operations performed on it during the
     * pairing algorithm to optimize pairing computations.
     */
    virtual bb::pairing::miller_lines const* get_precomputed_g2_lines() const = 0;
    /**
     *  @brief Returns the first G_1 element from the CRS, used by the Shplonk verifier to compute the final
     * commtiment.
     */
    virtual Curve::AffineElement get_first_g1() const = 0;
};

template <> class VerifierCrs<curve::Grumpkin> {
    using Curve = curve::Grumpkin;

  public:
    /**
     * @brief Returns the G_1 elements in the CRS after the pippenger point table has been applied on them
     *
     */
    virtual Curve::AffineElement* get_monomial_points() const = 0;
    virtual size_t get_monomial_size() const = 0;
    /**
     * @brief Returns the first G_1 element from the CRS, used by the Shplonk verifier to compute the final
     * commtiment.
     */
    virtual Curve::AffineElement get_first_g1() const = 0;
};

/**
 * A factory class to return the prover crs and verifier crs on request.
 * You can construct an empty placeholder factory, because composers need to be given a factory at construction time.
 */
template <typename Curve> class CrsFactory {
  public:
    CrsFactory() = default;
    CrsFactory(CrsFactory&& other) = default;
    virtual ~CrsFactory() = default;
    virtual std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> get_prover_crs(size_t) { return nullptr; }
    virtual std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> get_verifier_crs([[maybe_unused]] size_t degree = 0)
    {
        return nullptr;
    }
};

} // namespace bb::srs::factories