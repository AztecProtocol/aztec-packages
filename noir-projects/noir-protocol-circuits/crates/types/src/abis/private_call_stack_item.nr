use crate::{
    abis::{function_data::FunctionData, private_circuit_public_inputs::PrivateCircuitPublicInputs},
    address::AztecAddress, constants::PRIVATE_CALL_STACK_ITEM_LENGTH,
    traits::{Deserialize, Serialize, Empty}, utils::reader::Reader
};

struct PrivateCallStackItem {
    // This is the _actual_ contract address relating to where this function's code resides in the
    // contract tree. Regardless of whether this is a call or delegatecall, this
    // `contract_address` _does not change_. Amongst other things, it's used as a lookup for
    // getting the correct code from the tree. There is a separate `storage_contract_address`
    // within a CallStackItem which varies depending on whether this is a call or delegatecall.
    contract_address: AztecAddress,
    function_data: FunctionData,
    public_inputs: PrivateCircuitPublicInputs,
}

impl Eq for PrivateCallStackItem {
    fn eq(self, other: Self) -> bool {
        self.contract_address.eq(other.contract_address) &
        self.function_data.eq(other.function_data) &
        self.public_inputs.eq(other.public_inputs)
    }
}

impl Serialize<PRIVATE_CALL_STACK_ITEM_LENGTH> for PrivateCallStackItem {
    fn serialize(self) -> [Field; PRIVATE_CALL_STACK_ITEM_LENGTH] {
        let mut fields: BoundedVec<Field, PRIVATE_CALL_STACK_ITEM_LENGTH> = BoundedVec::new();

        fields.push(self.contract_address.to_field());
        fields.extend_from_array(self.function_data.serialize());
        fields.extend_from_array(self.public_inputs.serialize());

        assert_eq(fields.len(), PRIVATE_CALL_STACK_ITEM_LENGTH);

        fields.storage
    }
}

impl Deserialize<PRIVATE_CALL_STACK_ITEM_LENGTH> for PrivateCallStackItem {
    fn deserialize(serialized: [Field; PRIVATE_CALL_STACK_ITEM_LENGTH]) -> Self {
        // TODO(#4390): This should accept a reader ^ to avoid copying data.
        let mut reader = Reader::new(serialized);

        let item = Self {
            contract_address: reader.read_struct(AztecAddress::deserialize),
            function_data: reader.read_struct(FunctionData::deserialize),
            public_inputs: reader.read_struct(PrivateCircuitPublicInputs::deserialize),
        };

        reader.finish();
        item
    }
}

impl Empty for PrivateCallStackItem {
    fn empty() -> Self {
        PrivateCallStackItem {
            contract_address: AztecAddress::empty(),
            function_data: FunctionData::empty(),
            public_inputs: PrivateCircuitPublicInputs::empty(),
        }
    }
}

#[test]
fn serialization_of_empty() {
    let item = PrivateCallStackItem::empty();
    let serialized = item.serialize();
    let deserialized = PrivateCallStackItem::deserialize(serialized);
    assert(item.eq(deserialized));
}
