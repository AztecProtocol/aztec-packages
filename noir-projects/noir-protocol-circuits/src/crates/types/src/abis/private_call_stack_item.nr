use crate::{
    abis::{function_data::FunctionData, private_circuit_public_inputs::PrivateCircuitPublicInputs},
    address::AztecAddress,
    constants::{GENERATOR_INDEX__CALL_STACK_ITEM, PRIVATE_CALL_STACK_ITEM_LENGTH}, hash::pedersen_hash,
    traits::{Deserialize, Hash, Serialize}, utils::reader::Reader
};

struct PrivateCallStackItem {
    // This is the _actual_ contract address relating to where this function's code resides in the
    // contract tree. Regardless of whether this is a call or delegatecall, this
    // `contract_address` _does not change_. Amongst other things, it's used as a lookup for
    // getting the correct code from the tree. There is a separate `storage_contract_address`
    // within a CallStackItem which varies depending on whether this is a call or delegatecall.
    contract_address: AztecAddress,
    function_data: FunctionData,
    public_inputs: PrivateCircuitPublicInputs,
}

impl Eq for PrivateCallStackItem {
    fn eq(self, other: Self) -> bool {
        self.contract_address.eq(other.contract_address) &
        self.function_data.eq(other.function_data) &
        self.public_inputs.eq(other.public_inputs)
    }
}

impl Serialize<PRIVATE_CALL_STACK_ITEM_LENGTH> for PrivateCallStackItem {
    fn serialize(self) -> [Field; PRIVATE_CALL_STACK_ITEM_LENGTH] {
        let mut fields: BoundedVec<Field, PRIVATE_CALL_STACK_ITEM_LENGTH> = BoundedVec::new(0); 

        fields.push(self.contract_address.to_field());
        fields.extend_from_array(self.function_data.serialize());
        fields.extend_from_array(self.public_inputs.serialize());

        assert_eq(fields.len(), PRIVATE_CALL_STACK_ITEM_LENGTH);

        fields.storage
    }
}

impl Deserialize<PRIVATE_CALL_STACK_ITEM_LENGTH> for PrivateCallStackItem {
    fn deserialize(serialized: [Field; PRIVATE_CALL_STACK_ITEM_LENGTH]) -> Self {
        // TODO(#4390): This should accept a reader ^ to avoid copying data.
        let mut reader = Reader::new(serialized);

        let item = Self {
            contract_address: reader.read_struct(AztecAddress::deserialize),
            function_data: reader.read_struct(FunctionData::deserialize),
            public_inputs: reader.read_struct(PrivateCircuitPublicInputs::deserialize),
        };

        reader.finish();
        item
    }
}

impl Hash for PrivateCallStackItem {
    fn hash(self) -> Field {
        pedersen_hash(self.serialize(), GENERATOR_INDEX__CALL_STACK_ITEM)
    }
}

#[test]
fn serialization_of_empty() {
    let item: PrivateCallStackItem = dep::std::unsafe::zeroed();
    let serialized = item.serialize();
    let deserialized = PrivateCallStackItem::deserialize(serialized);
    assert(item.eq(deserialized));
}

#[test]
fn empty_hash() {
    let mut item: PrivateCallStackItem = dep::std::unsafe::zeroed();
    item.function_data.is_private = true;
    let hash = item.hash();

    // Value from private_call_stack_item.test.ts "computes empty item hash" test
    assert_eq(hash, 0x07151bc7440eb0b1c4c61a0bcfacca82cf2aed402c20d3742557179f7fedb635);
}
