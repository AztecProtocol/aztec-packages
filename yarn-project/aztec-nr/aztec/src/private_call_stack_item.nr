use crate::abi::FunctionData;
use crate::abi::PrivateCircuitPublicInputs;
use crate::constants_gen::GENERATOR_INDEX__CALL_STACK_ITEM;
use crate::hash::pedersen_hash;

struct PrivateCallStackItem {
    contract_address: Field,
    function_data: FunctionData,
    public_inputs: PrivateCircuitPublicInputs,
    is_execution_request: bool,
}

impl PrivateCallStackItem {
    pub fn hash(self) -> Field {
        pedersen_hash([
            self.contract_address,
            self.function_data.hash(),
            self.public_inputs.hash(),
        ], GENERATOR_INDEX__CALL_STACK_ITEM)
    }
}
