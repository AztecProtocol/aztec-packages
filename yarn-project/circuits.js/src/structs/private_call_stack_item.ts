import { AztecAddress } from '@aztec/foundation/aztec-address';
import { pedersenHash } from '@aztec/foundation/crypto';
import { Fr } from '@aztec/foundation/fields';
import { BufferReader, FieldReader, serializeToBuffer, serializeToFields } from '@aztec/foundation/serialize';
import { FieldsOf } from '@aztec/foundation/types';

import { GeneratorIndex, PRIVATE_CALL_STACK_ITEM_LENGTH } from '../constants.gen.js';
import { CallContext } from './call_context.js';
import { CallRequest, CallerContext } from './call_request.js';
import { FunctionData } from './function_data.js';
import { PrivateCircuitPublicInputs } from './private_circuit_public_inputs.js';

/**
 * Call stack item on a private call.
 */
export class PrivateCallStackItem {
  constructor(
    /**
     * Address of the contract on which the function is invoked.
     */
    public contractAddress: AztecAddress,
    /**
     * Data identifying the function being called.
     */
    public functionData: FunctionData,
    /**
     * Public inputs to the private kernel circuit.
     */
    public publicInputs: PrivateCircuitPublicInputs,
  ) {}

  static getFields(fields: FieldsOf<PrivateCallStackItem>) {
    return [fields.contractAddress, fields.functionData, fields.publicInputs] as const;
  }

  toBuffer() {
    return serializeToBuffer(...PrivateCallStackItem.getFields(this));
  }

  toFields(): Fr[] {
    const fields = serializeToFields(...PrivateCallStackItem.getFields(this));
    if (fields.length !== PRIVATE_CALL_STACK_ITEM_LENGTH) {
      throw new Error(
        `Invalid number of fields for PrivateCallStackItem. Expected ${PRIVATE_CALL_STACK_ITEM_LENGTH}, got ${fields.length}`,
      );
    }
    return fields;
  }

  /**
   * Deserializes from a buffer or reader.
   * @param buffer - Buffer or reader to read from.
   * @returns The deserialized instance.
   */
  static fromBuffer(buffer: Buffer | BufferReader): PrivateCallStackItem {
    const reader = BufferReader.asReader(buffer);
    return new PrivateCallStackItem(
      reader.readObject(AztecAddress),
      reader.readObject(FunctionData),
      reader.readObject(PrivateCircuitPublicInputs),
    );
  }

  static fromFields(fields: Fr[] | FieldReader): PrivateCallStackItem {
    const reader = FieldReader.asReader(fields);
    return new PrivateCallStackItem(
      AztecAddress.fromFields(reader),
      FunctionData.fromFields(reader),
      PrivateCircuitPublicInputs.fromFields(reader),
    );
  }

  /**
   * Returns a new instance of PrivateCallStackItem with zero contract address, function data and public inputs.
   * @returns A new instance of PrivateCallStackItem with zero contract address, function data and public inputs.
   */
  public static empty(): PrivateCallStackItem {
    return new PrivateCallStackItem(
      AztecAddress.ZERO,
      FunctionData.empty({ isPrivate: true }),
      PrivateCircuitPublicInputs.empty(),
    );
  }

  isEmpty() {
    return this.contractAddress.isZero() && this.functionData.isEmpty() && this.publicInputs.isEmpty();
  }

  /**
   * Computes this call stack item hash.
   * @returns Hash.
   */
  public hash(): Fr {
    return pedersenHash(
      this.toFields().map(field => field.toBuffer()),
      GeneratorIndex.CALL_STACK_ITEM,
    );
  }

  /**
   * Creates a new CallRequest with values of the calling contract.
   * @returns A CallRequest instance with the contract address, caller context, and the hash of the call stack item.
   */
  public toCallRequest(parentCallContext: CallContext) {
    if (this.isEmpty()) {
      return CallRequest.empty();
    }

    const currentCallContext = this.publicInputs.callContext;
    const callerContext = currentCallContext.isDelegateCall
      ? new CallerContext(parentCallContext.msgSender, parentCallContext.storageContractAddress)
      : CallerContext.empty();
    return new CallRequest(
      this.hash(),
      parentCallContext.storageContractAddress,
      callerContext,
      new Fr(this.publicInputs.callContext.sideEffectCounter),
      this.publicInputs.endSideEffectCounter,
    );
  }
}
