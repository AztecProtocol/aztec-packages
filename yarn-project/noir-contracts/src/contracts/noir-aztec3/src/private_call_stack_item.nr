use crate::abi::CallContext;
use crate::abi::ContractDeploymentData;
use crate::context::PrivateFunctionContext;
use crate::abi::PrivateCircuitPublicInputs;
use crate::abi::CommitmentTreesRoots;
use crate::types::vec::BoundedVec;
use crate::utils::arr_copy_slice;

use crate::oracle::call_private_function::call_private_function_internal;


struct FunctionData {
    function_selector: Field,
    is_private: bool,
    is_constructor: bool,
}

impl FunctionData {
    fn hash(self) -> Field {
        dep::std::hash::pedersen([
            // TODO hash_index
            self.function_selector,
            self.is_private as Field,
            self.is_constructor as Field,
        ])[0]
    }
}

struct PrivateCallStackItem {
    contract_address: Field,
    function_data: FunctionData,
    public_inputs: PrivateCircuitPublicInputs,
    is_execution_request: bool,
}

impl PrivateCallStackItem {
    fn call(contract_address: Field, function_selector: Field, args: [Field; crate::abi::MAX_ARGS], mut context: PrivateFunctionContext) -> (Self, PrivateFunctionContext) {
        let fields = call_private_function_internal(
            contract_address, 
            function_selector, 
            args
        );
        let item = PrivateCallStackItem {
            contract_address: fields[0],
            function_data: FunctionData {
                function_selector: fields[1],
                is_private: fields[2] as bool,
                is_constructor: fields[3] as bool,
            },
            public_inputs: PrivateCircuitPublicInputs {
                call_context: CallContext {
                    is_contract_deployment: fields[4] as bool,
                    is_delegate_call : fields[5] as bool,
                    is_static_call : fields[6] as bool,

                    msg_sender : fields[7],
                    storage_contract_address : fields[8],
                    portal_contract_address : fields[9],
                },
                // TODO handle the offsets as a variable incremented during extraction?
                args_hash: fields[10],
                return_values: arr_copy_slice(fields, [0; crate::abi::MAX_RETURN_VALUES], 11),
                new_commitments: arr_copy_slice(fields, [0; crate::abi::MAX_NEW_COMMITMENTS], 15),
                new_nullifiers: arr_copy_slice(fields, [0; crate::abi::MAX_NEW_NULLIFIERS], 19),
                private_call_stack: arr_copy_slice(fields, [0; crate::abi::MAX_PRIVATE_CALL_STACK], 23),
                public_call_stack: arr_copy_slice(fields, [0; crate::abi::MAX_PUBLIC_CALL_STACK], 27),
                new_l2_to_l1_msgs:arr_copy_slice(fields, [0; crate::abi::MAX_L2_TO_L1_MSGS], 31),
                // TODO #588, relevant issue: https://github.com/AztecProtocol/aztec-packages/issues/588
                // encrypted_logs_hash:arr_copy_slice(fields, [0; 2], 44),
                // unencrypted_logs_hash:arr_copy_slice(fields, [0; 2], 46),
                // encrypted_log_preimages_length: fields[47],
                // unencrypted_log_preimages_length: fields[48],
                commitment_trees_roots: CommitmentTreesRoots {
                    // Must match order in `private_circuit_public_inputs.hpp`
                    private_data_tree_root : fields[33],
                    nullifier_tree_root : fields[34],
                    contract_tree_root : fields[35],
                    l1_to_l2_messages_tree_root : fields[36],
                },
                contract_deployment_data: ContractDeploymentData {
                    constructor_vk_hash : fields[37],
                    function_tree_root : fields[38],
                    contract_address_salt : fields[39],
                    portal_contract_address : fields[40],
                },
            },
            is_execution_request: fields[41] as bool,
        };
        constrain contract_address == item.contract_address;
        constrain function_selector == item.function_data.function_selector;

        // TODO hash_index
        // let args_hash = dep::std::hash::pedersen(args)[0];
        // constrain args_hash == item.public_inputs.args_hash;

        constrain item.is_execution_request == false;

        constrain item.public_inputs.call_context.is_delegate_call == false;
        constrain item.public_inputs.call_context.is_static_call == false;
        constrain item.public_inputs.call_context.is_contract_deployment == false;

        context.private_call_stack = context.private_call_stack.push(item.hash());

        (item, context)
    }

    fn hash(self) -> Field {
        dep::std::hash::pedersen([
            // TODO hash_index
            // FIXME - https://github.com/AztecProtocol/aztec-packages/issues/499
            self.contract_address,
            self.function_data.hash(),
            self.public_inputs.hash(),
        ])[0]
    }
}