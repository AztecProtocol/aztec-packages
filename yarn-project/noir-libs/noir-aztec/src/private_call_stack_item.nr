use crate::abi::CallContext;
use crate::abi::hash_args;
use crate::abi::ContractDeploymentData;
use crate::context::Context;
use crate::abi::PrivateCircuitPublicInputs;
use crate::abi::CommitmentTreesRoots;
use crate::types::vec::BoundedVec;
use crate::types::point::Point;
use crate::utils::arr_copy_slice;
use crate::constants_gen;
use crate::constants_gen::GENERATOR_INDEX__FUNCTION_DATA;
use crate::constants_gen::GENERATOR_INDEX__CALL_STACK_ITEM;

use crate::oracle::call_private_function::call_private_function_internal;
use crate::oracle::arguments;

struct FunctionData {
    function_selector: Field,
    is_private: bool,
    is_constructor: bool,
}

impl FunctionData {
    fn hash(self) -> Field {
        dep::std::hash::pedersen_with_separator([
            self.function_selector,
            self.is_private as Field,
            self.is_constructor as Field,
        ], GENERATOR_INDEX__FUNCTION_DATA)[0]
    }
}

struct PrivateCallStackItem {
    contract_address: Field,
    function_data: FunctionData,
    public_inputs: PrivateCircuitPublicInputs,
    is_execution_request: bool,
}

impl PrivateCallStackItem {
    fn call<ARGS_COUNT>(
        contract_address: Field, 
        function_selector: Field, 
        args: [Field; ARGS_COUNT], 
        mut context: Context) -> (Self, Context) {
        let args_hash = hash_args(args);
        assert(args_hash == arguments::pack_arguments(args));
        PrivateCallStackItem::call_with_packed_args(contract_address, function_selector, args_hash, context)
    }

    fn call_with_packed_args(contract_address: Field, function_selector: Field, args_hash: Field, mut context: Context) -> (Self, Context) {
        let fields = call_private_function_internal(
            contract_address, 
            function_selector, 
            args_hash
        );
        let item = PrivateCallStackItem {
            contract_address: fields[0],
            function_data: FunctionData {
                function_selector: fields[1],
                is_private: fields[2] as bool,
                is_constructor: fields[3] as bool,
            },
            public_inputs: PrivateCircuitPublicInputs {
                call_context: CallContext {
                    msg_sender : fields[4],
                    storage_contract_address : fields[5],
                    portal_contract_address : fields[6],
                    
                    is_delegate_call : fields[7] as bool,
                    is_static_call : fields[8] as bool,
                    is_contract_deployment: fields[9] as bool,
                },
                // TODO handle the offsets as a variable incremented during extraction?
                args_hash: fields[10],
                return_values: arr_copy_slice(fields, [0; constants_gen::RETURN_VALUES_LENGTH], 11),
                read_requests: arr_copy_slice(fields, [0; constants_gen::MAX_READ_REQUESTS_PER_CALL], 15),
                new_commitments: arr_copy_slice(fields, [0; constants_gen::MAX_NEW_COMMITMENTS_PER_CALL], 19),
                new_nullifiers: arr_copy_slice(fields, [0; constants_gen::MAX_NEW_NULLIFIERS_PER_CALL], 23),
                nullified_commitments: arr_copy_slice(fields, [0; constants_gen::MAX_NEW_NULLIFIERS_PER_CALL], 27),
                private_call_stack: arr_copy_slice(fields, [0; constants_gen::MAX_PRIVATE_CALL_STACK_LENGTH_PER_CALL], 31),
                public_call_stack: arr_copy_slice(fields, [0; constants_gen::MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL], 35),
                new_l2_to_l1_msgs: arr_copy_slice(fields, [0; constants_gen::MAX_NEW_L2_TO_L1_MSGS_PER_CALL], 39),
                encrypted_logs_hash: arr_copy_slice(fields, [0; constants_gen::NUM_FIELDS_PER_SHA256], 41),
                unencrypted_logs_hash: arr_copy_slice(fields, [0; constants_gen::NUM_FIELDS_PER_SHA256], 43),
                encrypted_log_preimages_length: fields[45],
                unencrypted_log_preimages_length: fields[46],
                commitment_trees_roots: CommitmentTreesRoots {
                    // Must match order in `private_circuit_public_inputs.hpp`
                    private_data_tree_root : fields[47],
                    nullifier_tree_root : fields[48],
                    contract_tree_root : fields[49],
                    l1_to_l2_messages_tree_root : fields[50],
                    blocks_tree_root : fields[51],
                },
                contract_deployment_data: ContractDeploymentData {
                    deployer_public_key: Point::new(fields[52], fields[53]),
                    constructor_vk_hash : fields[54],
                    function_tree_root : fields[55],
                    contract_address_salt : fields[56],
                    portal_contract_address : fields[57],
                },
                chain_id: fields[58],
                version: fields[59],
            },
            is_execution_request: fields[60] as bool,
        };
        assert(contract_address == item.contract_address);
        assert(function_selector == item.function_data.function_selector);

        assert(args_hash == item.public_inputs.args_hash);

        assert(item.is_execution_request == false);

        // Assert that the call context of the enqueued call generated by the oracle matches our request.
        // We are issuing a regular call which is not delegate, static, or deployment. We also constrain
        // the msg_sender in the nested call to be equal to our address, and the execution context address
        // for the nested call to be equal to the address we actually called.
        assert(item.public_inputs.call_context.is_delegate_call == false);
        assert(item.public_inputs.call_context.is_static_call == false);
        assert(item.public_inputs.call_context.is_contract_deployment == false);
        assert(item.public_inputs.call_context.msg_sender == context.inputs.call_context.storage_contract_address);
        assert(item.public_inputs.call_context.storage_contract_address == contract_address);

        context.private_call_stack = context.private_call_stack.push(item.hash());

        (item, context)
    }

    fn hash(self) -> Field {
        dep::std::hash::pedersen_with_separator([
            self.contract_address,
            self.function_data.hash(),
            self.public_inputs.hash(),
        ], GENERATOR_INDEX__CALL_STACK_ITEM)[0]
    }
}