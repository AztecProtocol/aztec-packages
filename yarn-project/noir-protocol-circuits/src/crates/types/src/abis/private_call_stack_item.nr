use crate::{
    abis::{
        function_data::FunctionData,
        private_circuit_public_inputs::PrivateCircuitPublicInputs,
    },
    address::AztecAddress,
    constants::{
        GENERATOR_INDEX__CALL_STACK_ITEM,
        PRIVATE_CALL_STACK_ITEM_LENGTH,
    },
    hash::pedersen_hash,
    traits::{
        Deserialize,
        Hash,
        Serialize,
    },
    utils::reader::Reader,
};

struct PrivateCallStackItem {
    // This is the _actual_ contract address relating to where this function's code resides in the
    // contract tree. Regardless of whether this is a call or delegatecall, this
    // `contract_address` _does not change_. Amongst other things, it's used as a lookup for
    // getting the correct code from the tree. There is a separate `storage_contract_address`
    // within a CallStackItem which varies depending on whether this is a call or delegatecall.
    contract_address: AztecAddress,
    function_data: FunctionData,
    public_inputs: PrivateCircuitPublicInputs,
    // Not really needed for PrivateCallStackItem.
    is_execution_request: bool,
}

impl Serialize<PRIVATE_CALL_STACK_ITEM_LENGTH> for PrivateCallStackItem {
    fn serialize(self) -> [Field; PRIVATE_CALL_STACK_ITEM_LENGTH] {
        let mut fields: BoundedVec<Field, PRIVATE_CALL_STACK_ITEM_LENGTH> = BoundedVec::new(0); 

        fields.push(self.contract_address.to_field());
        fields.extend_from_array(self.function_data.serialize());
        fields.extend_from_array(self.public_inputs.serialize());
        fields.push(self.is_execution_request as Field);

        assert_eq(fields.len(), PRIVATE_CALL_STACK_ITEM_LENGTH);

        fields.storage
    }
}

impl Deserialize<PRIVATE_CALL_STACK_ITEM_LENGTH> for PrivateCallStackItem {
    fn deserialize(serialized: [Field; PRIVATE_CALL_STACK_ITEM_LENGTH]) -> Self {
        // TODO(#4390): This should accept a reader ^ to avoid copying data.
        let mut reader = Reader::new(serialized);

        let item = Self {
            contract_address: reader.read_struct(AztecAddress::deserialize),
            function_data: reader.read_struct(FunctionData::deserialize),
            public_inputs: reader.read_struct(PrivateCircuitPublicInputs::deserialize),
            is_execution_request: reader.read() as bool,
        };

        reader.finish();
        item
    }
}

impl Hash for PrivateCallStackItem {
    fn hash(self) -> Field {
        pedersen_hash(self.serialize(), GENERATOR_INDEX__CALL_STACK_ITEM)
    }
}

#[test]
fn serialization_smoke() {
    let item: PrivateCallStackItem = dep::std::unsafe::zeroed();
    let _serialized = item.serialize();
}

#[test]
fn hash_smoke() {
    let item: PrivateCallStackItem = dep::std::unsafe::zeroed();
    let _hashed = item.hash();
}
