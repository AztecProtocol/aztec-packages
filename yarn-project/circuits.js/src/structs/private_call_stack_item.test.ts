import { PRIVATE_CALL_STACK_ITEM_LENGTH } from '../constants.gen.js';
import { makePrivateCallStackItem } from '../tests/factories.js';
import { PrivateCallStackItem } from './private_call_stack_item.js';

describe('PrivateCallStackItem', () => {
  let item: PrivateCallStackItem;

  beforeAll(() => {
    const randomInt = Math.floor(Math.random() * 1000);
    item = makePrivateCallStackItem(randomInt);
  });

  it('serializes to buffer and deserializes it back', () => {
    const buffer = item.toBuffer();
    const res = PrivateCallStackItem.fromBuffer(buffer);
    expect(res).toEqual(item);
  });

  it('serializes to field array and deserializes it back', () => {
    const fieldArray = item.toFields();
    const res = PrivateCallStackItem.fromFields(fieldArray);
    expect(res).toEqual(item);
  });

  it('number of fields matches constant', () => {
    const fields = item.toFields();
    expect(fields.length).toBe(PRIVATE_CALL_STACK_ITEM_LENGTH);
  });

  it('computes hash', () => {
    const seed = 9870243;
    const item = makePrivateCallStackItem(seed);
    const hash = item.hash();
    expect(hash).toMatchSnapshot();
  });

  it('computes empty item hash', () => {
    const item = PrivateCallStackItem.empty();
    const hash = item.hash();
    expect(hash).toMatchSnapshot();

    // Value used in empty_hash test in private_call_stack_item.nr
    // console.log("hash", hash.toString());
  });
});
