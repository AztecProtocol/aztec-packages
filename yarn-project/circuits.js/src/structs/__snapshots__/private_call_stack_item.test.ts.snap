// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateCallStackItem computes empty item hash 1`] = `
Fr {
  "asBigInt": 3203485447791239934114288301818827303172706148767225207074779624230959822389n,
  "asBuffer": {
    "data": [
      7,
      21,
      27,
      199,
      68,
      14,
      176,
      177,
      196,
      198,
      26,
      11,
      207,
      172,
      202,
      130,
      207,
      42,
      237,
      64,
      44,
      32,
      211,
      116,
      37,
      87,
      23,
      159,
      127,
      237,
      182,
      53,
    ],
    "type": "Buffer",
  },
}
`;

exports[`PrivateCallStackItem computes hash 1`] = `
Fr {
  "asBigInt": 19520755359745729990898807497202224879281656008242555374727476866221326621394n,
  "asBuffer": {
    "data": [
      43,
      40,
      91,
      34,
      118,
      222,
      136,
      132,
      51,
      136,
      163,
      211,
      132,
      203,
      127,
      229,
      225,
      28,
      122,
      184,
      120,
      133,
      89,
      26,
      226,
      241,
      118,
      82,
      53,
      74,
      90,
      210,
    ],
    "type": "Buffer",
  },
}
`;
